File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int max = 0;
        int start = 0;
        int end = 0;
        Set<Character> set = new HashSet<>();
        while (end < s.length()) {
            char c = s.charAt(end);
            if (set.contains(c)) {
                set.remove(s.charAt(start++));
            } else {
                set.add(c);
                max = Math.max(max, set.size());
                end++;
            }
        }
        return max;
    }
}

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        result = 0
        _set = set()
        start = 0
        end = 0
        while end < len(s):
            c = s[end]
            if c in _set:
                _set.remove(s[start])
                start += 1
            else:
                _set.add(c)
                result = max(result, len(_set))
                end += 1
        return result